Implemented range setting in QNN llama flow

rohansjoshi opened the pull request:
Summary:
`llama.py` now has the `--range_setting` flag, for which there are the options `mse_weight_only` and `mse_with_act_loss`. There is also an eval script for computing perplexity called `eval_llama_qnn.py` (for faster eval, try seq length 1024). This script also has a flag --quant_linear_only to only quantize linear/conv nodes, to run faster experiments.

Commands:

```python examples/qualcomm/oss_scripts/llama/llama.py --checkpoint {MODEL_DIR}/consolidated.00.pth --params {MODEL_DIR}/params.json --tokenizer_path {MODEL_DIR}/tokenizer.model --max_seq_length 128 --ptq 16a4w --range_setting mse_with_act_loss```

```python examples/qualcomm/oss_scripts/llama/eval_llama_qnn.py --checkpoint {MODEL_DIR}/consolidated.00.pth --params {MODEL_DIR}/params.json --tokenizer_path {MODEL_DIR}/tokenizer.model --max_seq_length 128 --ptq 16a4w --range_setting mse_with_act_loss```

Update: I've also added SpinQuant as a feature to further improve accuracy. Both `llama.py` and `eval_llama_qnn.py` also have a flag `--spinquant` which can be used in combination with range setting or by itself. Based on my experiments on Llama 1B for 16a4w, I get the best results using both `--spinquant` and `--range_setting mse_with_act_loss`.

Rollback Plan:

Differential Revision: D78127727

cccclai commented on examples/qualcomm/oss_scripts/llama/llama.py:
```diff
@@ -597,6 +603,27 @@ def permute(w, heads):
+            model.ar_len = model.max_seq_len
+            tokens, atten_mask = model.get_example_inputs(use_kv_cache=False)
+            atten_mask.to(torch.float)
+            print(atten_mask.shape)
```
Removing debugging line

cccclai commented on examples/qualcomm/oss_scripts/llama/llama.py:
```diff
@@ -668,12 +696,12 @@ def permute(w, heads):
-                        kv_quant_attrs=kv_quant_attrs,
-                    ),
-                )
+                # custom_annotations = custom_annotations + (
```
Actually I need to have a separate PR for this.

cccclai reviewed (commented):
Still reading, will finish reading in a bit

cccclai reviewed (approved):
Thanks for improving the accuracy on llama + htp!

rohansjoshi committed:
Implemented range setting in QNN llama flow (#12377)

Summary:

`llama.py` now has the `--range_setting` flag, for which there are the options `mse_weight_only` and `mse_with_act_loss`. There is also an eval script for computing perplexity called `eval_llama_qnn.py` (for faster eval, try seq length 1024). This script also has a flag --quant_linear_only to only quantize linear/conv nodes, to run faster experiments.

Update: I've also added SpinQuant as a feature to further improve accuracy. Both `llama.py` and `eval_llama_qnn.py` also have a flag `--spinquant` which can be used in combination with range setting or by itself. Based on my experiments on Llama 1B, I get the best results using both `--spinquant` and `--range_setting mse_with_act_loss`.

Reviewed By: cccclai

Differential Revision: D78127727
